### PR TITLE
[SPARK-50573][SS] Adding State Schema ID to State Rows to schema evolution

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

In order to support schema evolution without a full rewrite, we are prepending each state row with a state schema ID, which we will use in subsequent changes to look up the schema that a row was written with. This change adds state schema IDs to both key and value rows in order to enable schema evolution when Avro is enabled.
For NoPrefixKeys and Values, the state schema ID is prepended to the row. 
For PrefixKeys and RangeKeys, the state schema ID is prepended to the remaining portion of the key (so the format is prefix key, schema ID, remaining key)
### Why are the changes needed?

These changes are needed in order to support schema evolution when Avro is enabled.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added unit test in RocksDBSuite to test all 4 prefix behaviors.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:54]: can we add some comments here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:88]: hmm - why do we need to pass the provider here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:177]: nit: extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1390]: nit: intentional ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1426]: nit: extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:631]: maybe also add a note to explain how many schema changes we can allow at a maximum

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (anishshri-db) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:1713]: Can we do this for some of the existing tests too ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:153]: Could we move this to a base class to make testing easier ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:193]: should we move this to the testing code ?

**Review by anishshri-db — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:59]: I don't think so? it lines up with the var name right?

**Comment by ericm-db** [comment]
@HeartSaVioR PTAL

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:978]: nit: one empty line below

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:983]: Is this still used?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:992]: Shall we have a fast-path if neither useColumnFamilies nor supportSchemaEvolution are true? You can just return the same array, unless this method's semantic is to always produce "new array". 

Please check with the callers of this method to see the feasibility; if it's not feasible due to caller (caller requires new array), let's rather add method doc to clarify that it provide a new byte array.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:956]: Wait... I see an issue with this.

When we perform prefix scan or range scan, the result should not be bound to the same schema ID. When performing prefix scan, Spark should be able to scan data from multiple schema IDs.

That said, the data layout should be following:

```
[vcf ID (optional] [prefix/range data (optional)] [schema ID (optional)] [data]
```

Unfortunately this is going to break the abstraction we have made. 
(The assumption here is that both `vcfId` and `schemaId` can be handled separately from data part including prefix/range. It does not seem to be true.)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:1818]: No, this does not cover the changes exhaustively.

We have to have tests which perform prefix scan and range scan "with evolving schema" among data, and see whether prefix (grouping) and range (ordering) do not break among data in different schema IDs.

**Review by HeartSaVioR — COMMENTED** [review]
It looked like a great abstraction, but I figured out this abstraction does not work with prefix scan and range scan. Let's revisit the design. I've stopped reviewing the code as it won't work unless I'm missing something.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:956]: Okay - let me remove schema id from these encoders.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:983]: Yeah we do use it for prefix scan

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (ericm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:1818]: Addressed above - removing schema id from prefix and range scan key encoders.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:409]: Removed above

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:54]: nit: extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:65]: same here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:200]: Lets mention order in which the fields are written or some ascii art maybe ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1073]: schema id is included here right ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:699]: Where is this used ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:55]: Lets add class comments here too

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:996]: @ericm-db - can u update this thread based on our discussion ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1073]: This is just prefix encoding - we are encoding the schema ID in the data portion (of just the Avro encoder)

**Comment by anishshri-db** [comment]
LGTM overall pending CI

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1260]: nit: unnecessary empty line

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1061]: nit: This might be worth keeping as it is.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1470]: nit: obsolete? schema ID should be a part of data

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:730]: It looks like we have a case of missing schema ID when the schema of prefix/range scan is the same with key schema. Do I understand correctly, or if we address this somewhere, could you please point it out?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:307]: nit: new line between L306 and L307 (here)

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala:363]: nit: do we need additional bracket while we also use withClue with code block? what's the usage? same for others.

**Review by HeartSaVioR — COMMENTED** [review]
Do we have an end-to-end test where Avro is enabled? If it's not available, can we have some tests which perform StateStore operations including prefix scan & range scan with Avro, so that we are not missing something? I don't see such tests in code change; please correct me if I'm missing something.

**Comment by ericm-db** [comment]
>Do we have an end-to-end test where Avro is enabled? If it's not available, can we have some tests which perform StateStore operations including prefix scan & range scan with Avro, so that we are not missing something? I don't see such tests in code change; please correct me if I'm missing something.

Yeah, all of our TWS Suites run with Avro enabled, I think this should be sufficient

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1061]: Do you have a strong preference here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:729]: Added a test case for this

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1061]: No I see it's now widely used, so we can't put this everywhere. No strong voice.

Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala:1061]: If it's okay with you I think I'll keep it as is

**Review by HeartSaVioR — APPROVED** [review]
+1 except PR description. It's a bit outdated. 
@ericm-db could you please update the description to fit to the new approach? Thanks!

**Comment by HeartSaVioR** [comment]
Pending CI. Once CI is green I'll merge.

**Comment by HeartSaVioR** [comment]
https://github.com/ericm-db/spark/runs/34898167412

CI is green. Thanks! Merging to master.

**Pull request closed without merging**
